pkg/controllers/nodeoverlay/controller.go

```diff
@@ -0,0 +1,256 @@
+		}
+		nodePoolToInstanceTypes[np.Name] = its
+	}
+
+	overlayList.OrderByWeight()
+	for i := range overlayList.Items {
+		if err := overlayList.Items[i].RuntimeValidate(ctx); err != nil {
+			overlayWithRuntimeValidationFailure[overlayList.Items[i].Name] = err
+			continue
+		}
+
+		// Due to reserved capacity type offering being dynamically injected as part of the GetInstanceTypes call
+		// We will need to make sure we are validating against each nodepool to make sure. This will ensure that
+		// overlays that are targeting reserved instance offerings will be able to apply the offering.
+		for _, np := range nodePoolList.Items {
+			overlaysWithConflict = append(overlaysWithConflict, c.checkOverlayPerNodePool(np, nodePoolToInstanceTypes[np.Name], overlayList.Items[i])...)
+		}
+	}
+
+	if err := c.updateOverlayStatuses(ctx, overlayList.Items, overlaysWithConflict, overlayWithRuntimeValidationFailure); err != nil {
+		return reconcile.Result{}, fmt.Errorf("updating nodeoverlay statuses, %w", err)
+	}
+
+	c.instanceTypeStore.UpdateStore(c.temporaryStore.updates)
+	return reconcile.Result{RequeueAfter: 6 * time.Hour}, nil
+}
+
+func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+	b := controllerruntime.NewControllerManagedBy(m).
+		Named(c.Name()).
+		// The reconciled overlay does not matter in this case as one reconcile loop
+		// will compare every overlay against every other overlay in the cluster.
+		For(&v1alpha1.NodeOverlay{}).
+		Watches(&v1.NodePool{}, NodeOverlayEventHandler(c.kubeClient)).
+		WithEventFilter(predicate.GenerationChangedPredicate{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 1,
+			RateLimiter:             reasonable.RateLimiter(),
+		})
+
+	for _, nodeClass := range c.cloudProvider.GetSupportedNodeClasses() {
+		b.Watches(nodeClass, NodeOverlayEventHandler(c.kubeClient))
+	}
+
+	return b.Complete(c)
+}
+
+func (c *Controller) checkOverlayPerNodePool(nodePool v1.NodePool, its []*cloudprovider.InstanceType, overlay v1alpha1.NodeOverlay) []string {
+	overlaysWithConflict := sets.Set[string]{}
+	overlayRequirements := scheduling.NewNodeSelectorRequirements(overlay.Spec.Requirements...)
+
+	for _, it := range its {
+		offerings := getOverlaidOfferings(nodePool, it, overlayRequirements)
+		// if we are not able to find any offerings for an instance type
+		// This will mean that the overlay does not select on the instance all together
+		if len(offerings) == 0 {
+			continue
+		}
+
+		conflictingPriceOverlay := c.isPriceUpdatesConflicting(nodePool.Name, it.Name, offerings, overlay)
+		if len(conflictingPriceOverlay) == 0 {
+			c.temporaryStore.updateInstanceTypeOffering(nodePool.Name, it.Name, overlay, offerings)
+		}
+		conflictingCapacityOverlay := c.isCapacityUpdatesConflicting(nodePool.Name, it.Name, overlay)
+		if len(conflictingCapacityOverlay) == 0 {
+			c.temporaryStore.updateInstanceTypeCapacity(nodePool.Name, it.Name, overlay)
+		}
```

Comment:
We only want to perform the updates if **neither** of the overlays are in conflict, right? Otherwise we could end up in a state where we apply the pricing override for an overlay, not the capacity override, and mark the status as conflicting.

Reply:
Also, for the conflict detection algorithm we discussed to work, we need to make sure we're updating the lowest tracked cost for an update regardless of if there's a conflict.

Reply:
Combined all these function to all be applied if there are no conflicts